src/parser/av1_parser.cpp

```diff
@@ -145,20 +145,25 @@ ParserResult Av1VideoParser::ParsePictureData(const uint8_t *p_stream, uint32_t
                 return PARSER_INVALID_ARG;
             }
             if (pfn_display_picture_cb_) {
-                decode_buffer_pool_[dpb_buffer_.frame_store[disp_idx].dec_buf_idx].use_status |= kFrameUsedForDisplay;
+                if (seq_header_.film_grain_params_present && frame_header_.film_grain_params.apply_grain) {
+                    disp_idx = dpb_buffer_.frame_store[disp_idx].fg_buf_idx;
+                } else {
+                    disp_idx = dpb_buffer_.frame_store[disp_idx].dec_buf_idx;
+                }
+                decode_buffer_pool_[disp_idx].use_status |= kFrameUsedForDisplay;
                 // Insert into output/display picture list
                 if (num_output_pics_ >= dec_buf_pool_size_) {
                     ERR("Display list size larger than decode buffer pool size!");
                     return PARSER_OUT_OF_RANGE;
                 } else {
-                    output_pic_list_[num_output_pics_] = dpb_buffer_.frame_store[disp_idx].dec_buf_idx;
+                    output_pic_list_[num_output_pics_] = disp_idx;
                     num_output_pics_++;
                 }
             }
             if ((ret = DecodeFrameWrapup()) != PARSER_OK) {
                 return ret;
             }
-        } else if (tile_group_data_.tile_number && tile_group_data_.tg_end == tile_group_data_.num_tiles - 1) {
+        } else if (tile_group_data_.num_tiles_parsed && tile_group_data_.num_tiles_parsed == tile_group_data_.num_tiles) {
```

Comment:
why the need for the first condition?

Reply:
This is to take care of the case where the tile group data and the frame header are not in the same OBU. In this case, num_tiles_parsed = num_tiles = 0 after frame header OBU is parsed. We need to skip the decode call until we parse tile data to the frame end.